### PR TITLE
Authentication in HTTP client and server middleware

### PR DESCRIPTION
**Pull request by shakedfb**

### CODE REVIEW

File: v2/fb-xcom-api-validator/src/config/index.ts [hunk 1]
```diff
@@ -34,6 +34,49 @@ const config = convict({
       env: 'LOG_LEVEL',
     },
   },
+  auth: {
```
Comment (yanivmo) [inline v2/fb-xcom-api-validator/src/config/index.ts:37]: Call it `authentication` please.

File: v2/fb-xcom-api-validator/src/auth-provider.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    public static getInstance(): AuthProvider {
+        if (!AuthProvider.instance) {
+            AuthProvider.instance = new AuthProvider();
+        }
+        return AuthProvider.instance;
+    }
+
+    public getAuthHeaders(method: HTTPMethod, endpoint: string, body: any, timestamp?: number, nonce?: string): AuthHeaders {
```
Comment (yanivmo) [inline v2/fb-xcom-api-validator/src/auth-provider.ts:46]: Only key and signature could be called authentication headers.

**Review by yanivmo — CHANGES_REQUESTED** [review]

File: v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  if (!nonce) {
+    throw new MissingAuthHeaders("nonce");
+  }
+  if (!signature) {
+    throw new MissingAuthHeaders("signature");
+  }
+  if (!apiKey) {
+    throw new MissingAuthHeaders("key");
+  }
+}
```
[the remark this answers is not in the document]
Comment (shakedfb) [reply v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts:64]: Got it, will use fastify validation schema to do that

File: v2/fb-xcom-api-validator/src/client/http.ts [hunk 4]
```diff
@@ -10,6 +11,14 @@ export class Http {
     };
   }
 
+  private enrichSecurityHeaders(method: HTTPMethod, endpoint: string, body?: any, headers?: Record<string, string>): Record<string, string> {
```
[the remark this answers is not in the document]
Comment (shakedfb) [reply v2/fb-xcom-api-validator/src/client/http.ts:14]: Not sure I understand this one

File: v2/fb-xcom-api-validator/src/encoding.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import * as base32 from "hi-base32";
+
+export class UnsupportedEncodingFormatError extends Error {}
+
+export type Encoding = "plain" | "base64" | "hexstr" | "base32" | "base58";
+
+export abstract class Encoder {
+    public abstract encode(payload: string): string;
+    public abstract decode(payload: string): string;
+}
```
Comment (yanivmo) [inline v2/fb-xcom-api-validator/src/encoding.ts:11]: This should be an interface, not an abstract class. Abstract classes are very very rare in TS and it should be a warning for you, if you find yourself creating one.

File: v2/fb-xcom-api-validator/tests/encoding.test.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+describe("Encoding methods", () => {
+    const data = "All in the golden afternoon Full leisurely we glide";
+    const base64Encoded = "QWxsIGluIHRoZSBnb2xkZW4gYWZ0ZXJub29uIEZ1bGwgbGVpc3VyZWx5IHdlIGdsaWRl";
+    const hexEncoded = "416c6c20696e2074686520676f6c64656e2061667465726e6f6f6e2046756c6c206c6569737572656c7920776520676c696465";
+    const base32Encoded = "IFWGYIDJNYQHI2DFEBTW63DEMVXCAYLGORSXE3TPN5XCARTVNRWCA3DFNFZXK4TFNR4SA53FEBTWY2LEMU======";
+    const base58Encoded = "4ZMy2teLGsR5CW9yw1h1pBaJuc3wEPNJZ7h2t9vnJimLJjUhvwSc3FPFQXyJ2p1BTLXdMn";
+
+    describe("Plain", () => {
+        const encoder = new Plain();
+        const encoded = encoder.encode(data);
```
Comment (yanivmo) [inline v2/fb-xcom-api-validator/tests/encoding.test.ts:12]: In Jest test, compute things always inside `it`, `beforeAll` or `beforeEach`. Never directly inside `describe`.

File: v2/fb-xcom-api-validator/tests/encoding.test.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        it("Encoded payload should match the input", () => {
+            expect(encoded).toBe(data);
+        })
+
+        it("Decoded payload should match the original input", () => {
```
Comment (yanivmo) [inline v2/fb-xcom-api-validator/tests/encoding.test.ts:17]: `it` description should always start with lower-case `should`.

File: v2/fb-xcom-api-validator/src/auth-provider.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        nonce = randomUUID();
+    }
+    if (!timestamp) {
+        timestamp = Date.now();
+    }
+    const signature = getRequestSignature(method, endpoint, body, timestamp, nonce);
+
+    return {
+        "X-FBAPI-KEY": authConfig.apiKey,
+        "X-FBAPI-SIGNATURE": encodeURIComponent(signature),
```
Comment (yanivmo) [inline v2/fb-xcom-api-validator/src/auth-provider.ts]: I'd prefer all the config values to be supplied here to `getRequestSignature` and not be buried deep inside the call stack — it makes testing more difficult.

File: v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
+import { FastifyReply, FastifyRequest, HookHandlerDoneFunction } from 'fastify';
+import logger from '../../logging';
+import { HTTPMethod, verifySignature } from '../../auth-provider';
+import { InvalidSignatureError } from '../../signing';
+
+const log = logger('handler:auth');
```
Comment (yanivmo) [inline v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts:6]: It's not a handler.

File: v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+  done();
+}
+
+function getAuthHeaders(headers: string[]) {
+  const getHeader = (headerName: string): string => {
+    const headerNameIndex = headers.findIndex(header => header === headerName);
+    const headerValueIndex = headerNameIndex + 1
+    return headers[headerValueIndex];
+  }
```
Comment (yanivmo) [inline v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts:34]: Are you sure it has to be so complicated? To my knowledge `request.headers` gives you very nice object.

File: v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import { FastifyReply, FastifyRequest, HookHandlerDoneFunction } from 'fastify';
+import logger from '../../logging';
+import { HTTPMethod, verifySignature } from '../../auth-provider';
+import { InvalidSignatureError } from '../../signing';
+
+const log = logger('handler:auth');
+
+export function authMiddleware(request: FastifyRequest, reply: FastifyReply, done: HookHandlerDoneFunction) {
```
Comment (yanivmo) [inline v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts:8]: We will need a separate middleware for each header component.

File: v2/fb-xcom-api-validator/src/signing.ts [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+type HashingAlgorithm = "sha256" | "sha512" | "sha3-256"
+type SigningAlgorithm = "hmac" | "rsa" | "ecdsa"
+type NamedCurve = "prime256v1" | "secp256k1"
+type Options = { algorithm: HashingAlgorithm, curve?: NamedCurve }
+
+export abstract class Signer {
+    public abstract sign(payload: string, key: string, options: Options): string;
+    public abstract verify(payload: string, key: string, signature: string, options: Options): void;
+}
```
Comment (yanivmo) [inline v2/fb-xcom-api-validator/src/signing.ts:20]: Should be an interface.

File: v2/fb-xcom-api-validator/src/signing.ts [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+            errorCode: "invalid-signature",
+        }
+    }
+}
+export class AlgorithmNotSupportedError extends Error {}
+
+type HashingAlgorithm = "sha256" | "sha512" | "sha3-256"
+type SigningAlgorithm = "hmac" | "rsa" | "ecdsa"
+type NamedCurve = "prime256v1" | "secp256k1"
+type Options = { algorithm: HashingAlgorithm, curve?: NamedCurve }
```
Comment (yanivmo) [inline v2/fb-xcom-api-validator/src/signing.ts:15]: Didn't we say we don't need the curve type?

Comment (yanivmo) [inline v2/fb-xcom-api-validator/src/signing.ts:15]: `algorithm` here is very misleading. `hashAlgorithm`?

**Review by yanivmo — CHANGES_REQUESTED** [review]
Move `auth-provider.ts`, `encoding.ts` and `signing.ts` into their own directory. `src/security`?

File: v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts (hunk 10 above)
Comment (shakedfb) [reply v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts]: I did so this way due to fastify lowercasing the incoming headers so I wanted to make sure it was received uppercased as it should
Now that I validated it I feel safe using `request.headers` instead

File: v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts (hunk 11 above)
Comment (shakedfb) [reply v2/fb-xcom-api-validator/src/server/middlewares/auth-middleware.ts]: renamed to `verifySignatureMiddleware`

File: v2/fb-xcom-api-validator/src/signing.ts (hunk 13 above)
Comment (shakedfb) [reply v2/fb-xcom-api-validator/src/signing.ts]: Agreed, as suggested, now using `hashAlgorithm`

**Review by yanivmo — APPROVED** [review]

**Pull request merged**